runtime/sema/type_tags.go

```diff
@@ -624,6 +639,16 @@ func findSuperTypeFromUpperMask(joinedTypeTag TypeTag, types []Type) Type {
 		restrictedTypeMask,
 		transactionTypeMask:
 		return getSuperTypeOfDerivedTypes(types)
+	case anyResourceAttachmentMask:
+		if allTypesAreAttachments(types) {
+			return AnyResourceAttachmentType
+		}
+		return commonSuperTypeOfHeterogeneousTypes(types)
+	case anyStructAttachmentMask:
+		if allTypesAreAttachments(types) {
+			return AnyStructAttachmentType
+		}
+		return commonSuperTypeOfHeterogeneousTypes(types)
```

Comment:
cc @SupunS i wasn't 100% sure if this is correct, can you please take a look and let me know?

Reply:
Does each attachment have its own type? or do they all have one type:  `AnyResourceAttachmentType`/ `anyStructAttachmentType`?
If they all have one type, then can simply return that from here. e.g:
```go
case anyResourceAttachmentMask:
	return AnyResourceAttachmentType
case anyStructAttachmentMask:
	return anyStructAttachmentType
```

(Explanation: here if the `joinedTypeTag` is `anyStructAttachmentMask`, that means all the types are struct attachments, so can return that as the supertype. If there are mixed types, then `joinedTypeTag` would be different hence will exit from default-case, and will eventually go to `commonSuperTypeOfHeterogeneousTypes()`)

Reply:
In specifically the `AnyResource` case, because `anyResourceMask` is only a lower tag and `anyResourceAttachmentTag`, the switch over just the upper tag here will cause us to hit this case even if the two types that created the tag are `AnyResource` and `AnyResourceAttachment`. Hence we need to actually verify that all of the types in the list are actually attachments before we can say that the supertype is actually `AnyResourceAttachment`.

Reply:
ah yes, I noticed that too, it's actually a bug in the parent method and added a fix locally. Will send it today in a separate PR.

Reply:
Added in https://github.com/onflow/cadence/pull/2103. The two cases should be able to simplify as I suggested in the previous comment. let me know if still has problems.